bootstrap/src/basis/ioring.c

```diff
@@ -0,0 +1,554 @@
+        break;
+    default:
+        break;
+    }
+
+    return oe;
+}
+
+int
+hm_ioring_user_data_complete(hm_user_data_t *user_data, hm_ioring_t *ioring) {
+    hm_opt_error_t oe = HM_OE_NONE;
+
+    while (!hm_user_data_is_complete(user_data)) {
+        HM_OE(oe, hm_ioring_flush_cqes(1, ioring));
+    }
+
+OUT:
+    if (oe == HM_OE_NONE) {
+        // io_uring_cqe res returns -errno in the res field if there are errors. We do the same.
+        return user_data->cqe.res;
+    } else if (oe == HM_OE_ERROR) {
+        // Generic error, but this function returns -errno on failure. Just use the most appropriate
+        // system error.
+        return -EIO;
+    } else {
+        // oe was initialized from errno.
+        return -oe;
+    }
+}
+
+hm_user_data_t *
+hm_ioring_nop_submit(hm_ioring_t *ioring) {
+    hm_opt_error_t oe = HM_OE_NONE;
+    struct io_uring_sqe *sqe;
+    HM_OE(oe, hm_ioring_get_sqe(&sqe, ioring));
+
+    hm_user_data_t *user_data;
+    HM_OE(oe, hm_user_data_create(&user_data));
+    user_data->opcode = IORING_OP_NOP;
+
+    sqe->user_data = (uint64_t)user_data;
+    sqe->opcode = IORING_OP_NOP;
+
+OUT:
+    assert(oe == HM_OE_NONE);
+
+    return user_data;
+}
+
+hm_user_data_t *
+hm_ioring_open_submit(uint8_t *pathname, int flags, mode_t mode, hm_ioring_t *ioring) {
+    hm_opt_error_t oe = HM_OE_NONE;
+    struct io_uring_sqe *sqe;
+    HM_OE(oe, hm_ioring_get_sqe(&sqe, ioring));
+
+    hm_user_data_t *user_data;
+    HM_OE(oe, hm_user_data_create(&user_data));
+    user_data->opcode = IORING_OP_OPENAT;
+    user_data->pathname = pathname;
+
+    sqe->user_data = (uint64_t)user_data;
+    sqe->opcode = IORING_OP_OPENAT;
+    sqe->fd = AT_FDCWD;
+    sqe->addr = (uint64_t)pathname;
+    sqe->open_flags = flags;
+    sqe->len = mode;
+
+OUT:
+    assert(oe == HM_OE_NONE);
+
+    return user_data;
+}
+
+hm_user_data_t *
+hm_ioring_close_submit(int fd, hm_ioring_t *ioring) {
+    hm_opt_error_t oe = HM_OE_NONE;
+    struct io_uring_sqe *sqe;
+    HM_OE(oe, hm_ioring_get_sqe(&sqe, ioring));
+
+    hm_user_data_t *user_data;
+    HM_OE(oe, hm_user_data_create(&user_data));
+    user_data->opcode = IORING_OP_CLOSE;
+
+    sqe->user_data = (uint64_t)user_data;
+    sqe->opcode = IORING_OP_CLOSE;
+    sqe->fd = fd;
+
+OUT:
+    assert(oe == HM_OE_NONE);
+
+    return user_data;
+}
+
+hm_user_data_t *
+hm_ioring_read_submit(int fd, uint8_t *buffer, uint64_t n, hm_ioring_t *ioring) {
+    hm_opt_error_t oe = HM_OE_NONE;
+    struct io_uring_sqe *sqe;
+    HM_OE(oe, hm_ioring_get_sqe(&sqe, ioring));
+
+    hm_user_data_t *user_data;
+    HM_OE(oe, hm_user_data_create(&user_data));
+    user_data->opcode = IORING_OP_READ;
+    user_data->buffer = buffer;
+
+    sqe->user_data = (uint64_t)user_data;
+    sqe->opcode = IORING_OP_READ;
+    sqe->off = -1;  // Use current file position.
+    sqe->fd = fd;
+    sqe->addr = (uint64_t)buffer;
+    sqe->len = n;
+
+OUT:
+    assert(oe == HM_OE_NONE);
+
+    return user_data;
+}
```

Comment:
I'm not it's okay to just swallow a possible error in the C layer. We need to return a pointer (as `uns`) to OCaml so there's no way to return errors as negative values like in some of the other OCaml/C bindings.

That'd mean I'd need to return a `(sint * uns)` from C for the error and the user_data pointer. I'd also have to handle the error on the OCaml side. APIs like `File.Read.submit` would change slightly.

Reply:
I made the change I suggested here and will push it along with fixes from the current round of review.